File: tx_15.py
#!/usr/bin/env python3
import signal
import sys
import numpy as np
import SoapySDR
from SoapySDR import SOAPY_SDR_TX  # direction constants

# -------------------- Constants (match your C example) --------------------
CH                 = 0              # TX channel A
HOST_SR_HZ         = 5_000_000      # 5 Msps host-side
OVERSAMPLE         = 8              # target RF ≈ HOST_SR * 8 = 40 Msps (Lime internal)
TX_LPF_BW_HZ       = 50_000_000     # 50 MHz
LO_HZ              = 30_000_000     # 30 MHz LO
NCO_FREQ_HZ        = 15_000_000     # 15 MHz NCO magnitude
NCO_DOWNCONVERT    = True           # RF = LO - NCO  (so set BB = -NCO)
TX_GAIN_DB         = 40             # moderate TX gain
BUF_SAMPLES        = 8192           # chunk size (complex samples)
SEND_TIMEOUT_US    = 1_000_000      # 1 s
TONE_SCALE         = 0.70           # 70% full scale
STREAM_BUFFER_LEN  = 131072         # driver hint (samples); adjust if needed

keep_running = True
def on_sigint(signum, frame):
    global keep_running
    keep_running = False
signal.signal(signal.SIGINT, on_sigint)

# -------------------- Open first Lime device via Soapy --------------------
# If you have multiple SDRs, pass args like dict(driver="lime", serial="XXXX")
devs = SoapySDR.Device.enumerate(dict(driver="lime"))
if not devs:
    print("No LimeSDR found", file=sys.stderr)
    sys.exit(1)

sdr = SoapySDR.Device(devs[0])

# -------------------- Basic TX setup --------------------
# Enable TX channel (Soapy enables on stream activate, but we set params first)
sdr.setSampleRate(SOAPY_SDR_TX, CH, HOST_SR_HZ)
# (Lime oversampling is chosen internally based on requested host SR; aiming for ~40 Msps NCO range)

sdr.setBandwidth(SOAPY_SDR_TX, CH, TX_LPF_BW_HZ)
sdr.setGain(SOAPY_SDR_TX, CH, TX_GAIN_DB)

# RF LO at 30 MHz
sdr.setFrequency(SOAPY_SDR_TX, CH, "RF", LO_HZ)

# NCO (BB) at -15 MHz for down-convert so RF = LO - NCO = 15 MHz
bb = -NCO_FREQ_HZ if NCO_DOWNCONVERT else +NCO_FREQ_HZ
sdr.setFrequency(SOAPY_SDR_TX, CH, "BB", bb)

# Optional: printbacks (best-effort; not all drivers provide readback)
try:
    host_sr = sdr.getSampleRate(SOAPY_SDR_TX, CH)
    print(f"Set/Get: SampleRate host={host_sr/1e6:.2f} Msps")
except Exception:
    pass
try:
    print(f"Set/Get: TX Gain = {sdr.getGain(SOAPY_SDR_TX, CH):.0f} dB")
except Exception:
    pass
try:
    print(f"Set/Get: LO = {sdr.getFrequency(SOAPY_SDR_TX, CH, 'RF')/1e6:.6f} MHz")
except Exception:
    pass
# (Many drivers don’t expose an NCO index/readback; we skip it.)

# -------------------- TX stream --------------------
# Format 'CS16' = 16-bit IQ (matches LMS_FMT_I16)
st_args = {"bufferLength": STREAM_BUFFER_LEN}  # driver-specific hint; safe to omit
tx_stream = sdr.setupStream(SOAPY_SDR_TX, "CS16", [CH], st_args)
sdr.activateStream(tx_stream)  # start

print(f"TX stream started (bufferLength={STREAM_BUFFER_LEN}, fmt=CS16).")

# Build a constant IQ buffer: I=0.7 FS, Q=0 -> pure tone at NCO after up/downconversion
I = np.int16(TONE_SCALE * 32767)
Q = np.int16(0)
iq = np.empty(BUF_SAMPLES, dtype=np.complex64)  # build in float then view as int16 interleaved
iq.real = I
iq.imag = Q
# Convert to interleaved int16 (Soapy expects raw bytes)
iq_i16 = np.empty(2*BUF_SAMPLES, dtype=np.int16)
iq_i16[0::2] = I
iq_i16[1::2] = Q

print(f"TX @ {(LO_HZ - (NCO_FREQ_HZ if NCO_DOWNCONVERT else -NCO_FREQ_HZ))/1e6:.1f} MHz "
      f"(host={HOST_SR_HZ/1e6:.2f} Msps, gain={TX_GAIN_DB} dB). Ctrl+C to stop.")

# -------------------- Stream loop --------------------
while keep_running:
    # writeStream expects a pointer-like object; numpy gives a buffer interface
    sr = sdr.writeStream(tx_stream, [iq_i16], BUF_SAMPLES, timeoutUs=SEND_TIMEOUT_US)
    if sr.ret < 0:
        print(f"writeStream error: {sr.ret}", file=sys.stderr)
        break

print("\nSIGINT detected: muting TX and shutting down safely...")

# -------------------- Cleanup (graceful mute) --------------------
try:
    # push one zero buffer so last burst is silence
    z = np.zeros(2*BUF_SAMPLES, dtype=np.int16)
    _ = sdr.writeStream(tx_stream, [z], BUF_SAMPLES, timeoutUs=SEND_TIMEOUT_US)
except Exception:
    pass

try:
    sdr.deactivateStream(tx_stream)
    sdr.closeStream(tx_stream)
    print("TX stream stopped.")
except Exception:
    pass

File: tx_15.c
#include "lime/LimeSuite.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>
#include <stdbool.h>
#include <signal.h>

#define CH                0          // TX channel A
#define HOST_SR_HZ        5000000    // 5 Msps over USB
#define OVERSAMPLE        8          // RF ≈ HOST_SR * OVERSAMPLE = 40 Msps
#define TX_LPF_BW_HZ      50000000   // 50 MHz LPF
#define LO_HZ             30000000   // 30 MHz LO (PLL-friendly)
#define NCO_FREQ_HZ       15000000   // 15 MHz NCO magnitude
#define NCO_INDEX         0
#define NCO_DOWNCONVERT   true       // RF = LO - NCO = 15 MHz
#define TX_GAIN_DB        40         // moderate TX gain
#define FIFO_SIZE_SAMPLES (1<<17)    // bigger FIFO for stability
#define BUF_SAMPLES       8192       // larger chunk reduces IRQ/USB churn
#define SEND_TIMEOUT_MS   1000
#define TONE_SCALE        0.70       // 70% FS to avoid DAC clipping

#define CHECK(x) do { \
  int __e = (x); \
  if (__e) { fprintf(stderr,"ERROR: %s -> %s\n", #x, LMS_GetLastErrorMessage()); goto cleanup; } \
} while(0)

static volatile int keep_running = 1;
static void on_sigint(int s){ (void)s; keep_running = 0; }

static void print_sr(lms_device_t* dev){
    double host=0, rf=0;
    if (!LMS_GetSampleRate(dev, LMS_CH_TX, CH, &host, &rf))
        printf("Set/Get: SampleRate host=%.2f Msps, rf=%.2f Msps\n", host/1e6, rf/1e6);
}

static void print_lpf(lms_device_t* dev){
#ifdef LMS_GetLPFBW
    float_type bw = 0;
    if (!LMS_GetLPFBW(dev, LMS_CH_TX, CH, &bw))
        printf("Set/Get: TX LPF BW = %.2f MHz\n", bw/1e6);
#else
    printf("Note: LMS_GetLPFBW not available in this LimeSuite; skipping BW readback.\n");
#endif
}

static void print_gain(lms_device_t* dev){
    unsigned int g = 0;
    if (!LMS_GetGaindB(dev, LMS_CH_TX, CH, &g))
        printf("Set/Get: TX Gain = %u dB\n", g);
}

static void print_lo(lms_device_t* dev){
    double lof = 0;
    if (!LMS_GetLOFrequency(dev, LMS_CH_TX, CH, &lof))
        printf("Set/Get: LO = %.6f MHz\n", lof/1e6);
}

static void print_nco(lms_device_t* dev){
#ifdef LMS_GetNCOFrequency
    double freqs[16]={0};
    double ph=0.0;
    if (!LMS_GetNCOFrequency(dev, true, CH, freqs, &ph)) {
        int idx = LMS_GetNCOIndex(dev, true, CH);
        printf("Set/Get: NCO idx=%d, f=%.6f MHz, mode=%s\n",
               idx, freqs[idx]/1e6, (LMS_GetNCOIndex(dev, true, CH)>=0 && NCO_DOWNCONVERT)?"downconv":"upconv");
    }
#else
    int idx = LMS_GetNCOIndex(dev, true, CH);
    printf("Set/Get: NCO idx=%d (no frequency readback in this LimeSuite)\n", idx);
#endif
}

int main(void)
{
    lms_device_t* dev = NULL;
    lms_stream_t  txs;
    int16_t*      buf = NULL;

    signal(SIGINT, on_sigint);

    // 1) Open device
    lms_info_str_t list[8];
    int n = LMS_GetDeviceList(list);
    if (n < 1) { fprintf(stderr,"No LimeSDR found\n"); return 1; }
    if (LMS_Open(&dev, list[0], NULL)) { fprintf(stderr,"LMS_Open failed: %s\n", LMS_GetLastErrorMessage()); return 1; }

    // 2) Basic setup
    CHECK(LMS_Init(dev));
    CHECK(LMS_EnableChannel(dev, LMS_CH_TX, CH, true));
    printf("TX channel enabled.\n");

    // Sample rates: host 5 Msps, RF ≈ 40 Msps (for NCO range up to 20 MHz)
    CHECK(LMS_SetSampleRate(dev, HOST_SR_HZ, OVERSAMPLE));
    print_sr(dev);

    // TX LPF/BW
    CHECK(LMS_SetLPFBW(dev, LMS_CH_TX, CH, TX_LPF_BW_HZ));
    print_lpf(dev);

    // Gain
    CHECK(LMS_SetGaindB(dev, LMS_CH_TX, CH, TX_GAIN_DB));
    print_gain(dev);

    // LO
    CHECK(LMS_SetLOFrequency(dev, LMS_CH_TX, CH, LO_HZ));
    print_lo(dev);

    // // Calibrate near operating BW
    // CHECK(LMS_Calibrate(dev, LMS_CH_TX, CH, 20000000, 0)); // 20 MHz calib BW

    // 3) NCO: program 15 MHz and select downconvert so RF = 30 - 15 = 15 MHz
    {
        double freqs[16]={0};
        freqs[NCO_INDEX] = (double)NCO_FREQ_HZ; // positive; direction set below
        CHECK(LMS_SetNCOFrequency(dev, true, CH, freqs, 0.0));              // dir_tx=true
        CHECK(LMS_SetNCOIndex    (dev, true, CH, NCO_INDEX, NCO_DOWNCONVERT));
        int idx = LMS_GetNCOIndex(dev, true, CH);
        if (idx < 0) { fprintf(stderr,"LMS_GetNCOIndex failed: %s\n", LMS_GetLastErrorMessage()); goto cleanup; }
        print_nco(dev);
    }

    // 4) TX stream
    memset(&txs, 0, sizeof(txs));
    txs.channel  = CH;
    txs.isTx     = true;
    txs.fifoSize = FIFO_SIZE_SAMPLES;
    txs.dataFmt  = LMS_FMT_I16; // C-safe
    CHECK(LMS_SetupStream(dev, &txs));
    CHECK(LMS_StartStream(&txs));
    printf("TX stream started (fifo=%d samples, fmt=I16).\n", FIFO_SIZE_SAMPLES);

    // Constant DC IQ -> pure RF tone after NCO
    buf = (int16_t*)malloc(2*BUF_SAMPLES*sizeof(int16_t));
    if (!buf) { fprintf(stderr,"malloc failed\n"); goto cleanup; }
    const int16_t I = (int16_t)(TONE_SCALE * 32767.0);
    const int16_t Q = 0;
    for (size_t i=0;i<BUF_SAMPLES;i++){ buf[2*i+0]=I; buf[2*i+1]=Q; }

    double host_sr=0, rf_sr=0;
    LMS_GetSampleRate(dev, LMS_CH_TX, CH, &host_sr, &rf_sr); // best-effort
    unsigned int g_cur=0; LMS_GetGaindB(dev, LMS_CH_TX, CH, &g_cur);

    printf("TX @ %.1f MHz  (host=%.2f Msps, rf=%.2f Msps, gain=%u dB). Ctrl+C to stop.\n",
           (LO_HZ - NCO_FREQ_HZ)/1e6, host_sr/1e6, rf_sr/1e6, g_cur);

    // 5) Stream loop
    while (keep_running) {
        lms_stream_meta_t meta; memset(&meta,0,sizeof(meta));
        if (LMS_SendStream(&txs, buf, BUF_SAMPLES, &meta, SEND_TIMEOUT_MS) < 0) {
            fprintf(stderr,"LMS_SendStream error: %s\n", LMS_GetLastErrorMessage());
            break;
        }
    }

    printf("\nSIGINT detected: muting TX and shutting down safely...\n");

cleanup:
    // Graceful mute: push one zero buffer so the last burst is silence
    if (txs.handle) {
        int16_t* z = (int16_t*)calloc(2*BUF_SAMPLES, sizeof(int16_t));
        if (z){
            lms_stream_meta_t meta; memset(&meta,0,sizeof(meta));
            (void)LMS_SendStream(&txs, z, BUF_SAMPLES, &meta, SEND_TIMEOUT_MS);
            free(z);
        }
        LMS_StopStream(&txs);
        LMS_DestroyStream(dev, &txs);
        printf("TX stream stopped.\n");
    }

    // Disable TX channel (safe TX off)
    if (dev) {
        LMS_EnableChannel(dev, LMS_CH_TX, CH, false);
        printf("TX channel disabled.\n");
    }

    if (buf) free(buf);
    if (dev) LMS_Close(dev);
    return 0;
}
